### DISCUSSION TITLE
[Go] How to query for a function that is passed as a variable?

### DISCUSSION THREAD

**Discussion by JasperSurmont** [github/codeql#13159] [Q&A] [answered]
Example: 
```go
type A struct {
    Handler func(int)
}

var as = []A{
    {
        Handler: someFunc
    },
    {
        Handler: someFunc2
    },
    ...
}

for i := 0; i < len(as); i++{
    as[i].Handler(i)
}

```

I want the query to return all the functions that are being called by the `Handler`: `someFunc` , `someFunc2`, etc. How would I go about achieving this?

**Comment by JasperSurmont** [comment]
How I am currently solving this is by querying for all the `FunctionNames` that are the `value` in a `KeyValueExpr`. I then must match the `VariableName` to a predefined name (in my example this would be `as`). 

However, this method is not flexible at all: e.g. every change to structure of the variable `as` messes up this query. Iḿ hoping to get a bit of guidance on how I could create a more robust query.

**Reply by smowton** [reply]
From the example you give it looks like you're looking for interprocedural flow of specific handlers to callsites?

If so, how about something like

```ql
import go

class Config extends DataFlow::Configuration {
  Config() { this = "Global callee flow" }

  override predicate isSource(DataFlow::Node n) {
    n instanceof DataFlow::FunctionNode
  }

  predicate isUnknownCallee(DataFlow::CallNode cn, DataFlow::Node callee) {
    callee = cn.getCalleeNode() and
    not exists(cn.getACalleeIncludingExternals()) // Exclude callsites with a known callee
  }

  override predicate isSink(DataFlow::Node n) {
    this.isUnknownCallee(_, n)
  }
}

from Config c, DataFlow::Node source, DataFlow::Node sink, DataFlow::Node callNode
where c.hasFlow(source, sink) and
c.isUnknownCallee(callNode, sink)
select source, callNode
```

This works for me with the somewhat elaborated example

```go
package example

type A struct {
    Handler func(int)
}

func someFunc(x int) { }

func someFunc2(x int) { }

var as = []A{
    {
        Handler: someFunc,
    },
    {
        Handler: someFunc2,
    },
}

func user() {

    for i := 0; i < len(as); i++{
       as[i].Handler(i)
    }

}
```

**Reply by JasperSurmont** [reply]
Hey, thanks so much!

I indeed tested this on the example and it works. However, it doesn't work on my codebase, and I think the reason is because the place where the handlers are actually called is in an external library. I have read [this](https://codeql.github.com/docs/codeql-language-guides/modeling-data-flow-in-go-libraries/) but I can't seem to understand how I would be able to incorperate your solution with that. Would you mind giving me some directions?

**Reply by smowton** [reply]
What is the actual characteristic you're looking for? Flow of a function to a write to the field "Handler"?

**Reply by JasperSurmont** [reply]
I'm looking for the flow of a specified function (like in your elaborated example `user`) to any other function (also 'indirect' functions called like in the example). In my project, this specified function has other restrictions (e.g. no platform dependent variables like `int`), so I'd use this characteristic to then match several other.

However, with my limited knowledge I couldn't get it to work and thought the reason was the 'indirection' becsuse the source code never explicitly calls the function.

Maybe I just did something silly and it is really not difficult at all, for which my excuses.

**Reply by smowton** [reply]
You mentioned though that the actual call is happening in an opaque third-party library, so what is the sign that a particular function pointer will go on to be executed by that library? Is the function pointer assigned to a particular field? Passed to a library function?

Could you give another small example that is more representative of your actual problem?

**Reply by JasperSurmont** [reply]
Yes of course, sorry for the confusion.

An example can be found in [this]() repo, using the Cosmos SDK.

[Here](https://github.com/CosmosContracts/juno/blob/main/x/feeshare/types/tx.pb.go#L455) the variable `_Msg_serviceDesc` is 'registered'.  This variable  (declaration found [here](https://github.com/CosmosContracts/juno/blob/main/x/feeshare/types/tx.pb.go#L513)) contains Handlers which are function pointers. 

The [`RegisterMsgService`](https://github.com/cosmos/cosmos-sdk/blob/90e3e69b95a4ab0bd41fc523070f6a9df00d5b79/baseapp/msg_service_router.go#L54) basicallly just sets the key-value map of the name of the handler to the handler itself. 

Then, at one point, the Cosmos SDK calls [`runMsgs`](https://github.com/cosmos/cosmos-sdk/blob/90e3e69b95a4ab0bd41fc523070f6a9df00d5b79/baseapp/baseapp.go#L774), which will eventually do:
```go
handler := app.msgServiceRouter.Handler(msg)
if handler == nil {
	return nil, errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "can't route message %+v", msg)
}

msgResult, err := handler(ctx, msg)
```

What I want to achieve is querying all functions that are called when `runMsgs` is called, which thus also should include the `Handlers` passed to the `_Msg_serviceDesc` variable.

I hope this clears up my situation.
Thanks again for your time!

**Reply by smowton** [reply, marked as answer]
Depending on how precise you want to be, here's a sink that identifies any write to the field `MethodDesc.Handler`:

```ql
  override predicate isSink(DataFlow::Node n) {
    exists(Write w |
      w.writesField(_, any(Field f | f.hasQualifiedName("google.golang.org/grpc", "MethodDesc", "Handler")), n)
    )
  }
```

However perhaps you're concerned that not all MethodDesc.Handlers end up passed to RegisterService? In that case we can make our sink any argument to RegisterService:

```ql
  override predicate isSink(DataFlow::Node n) {
    n = any(DataFlow::CallNode cn |
      cn.getTarget().(Method).hasQualifiedName("github.com/gogo/protobuf/grpc", "Server", "RegisterService")
    ).getArgument(0)
  }
```

...then we must add additional flow steps to propagate flow from the Handler field to its enclosing MethodDesc, then to the array of MethodDescs, then to the ServiceDesc surrounding that. We can make array-writes taint-propagating (along with various other things; see the predicate `defaultAdditionalTaintStep`) by turning our `DataFlow::Configuration` into a `TaintTracking::Configuration`, then we can make field-writes taint-propagating by defining

```ql
  override predicate isAdditionalTaintStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
    exists(Write w |
      w.writesField(toNode.(DataFlow::PostUpdateNode).getPreUpdateNode(), _, fromNode)
    )
  }
```

That's pretty blunt -- we could restrict it to the Handler and Methods fields by replacing the `_` with an `any(Field f | ...)` definition like that used in our first sink definition if need be.

With those changes made analysing the `juno` database successfully finds that `_Query_FeeShares_Handler` and 19 other similarly-named handler functions make it to a RegisterService call.

**Reply by JasperSurmont** [reply]
Wow, thank you so much for your help! Exactly what I wanted to achieve!

**Reply by JasperSurmont** [reply]
Hey @smowton 

I have one additional question. So your solution works as expected (currently using your first suggestion without tainttracking), but I wanted to expand it such that a query finds all functions being called in the found `FunctionNode`. As an example, you mentioned you found `_Query_FeeShares_Handler`, but I'd like to go further and also find the functions being called in there, like [queryClient.FeeShare](https://github.com/CosmosContracts/juno/blob/main/x/feeshare/types/query.pb.go#L641), and recursively all the funcitons called in there. 

I came up with the following which kind of works, but I'm quite certain this is not a good/beautiful solution. Could you maybe guide me a little bit on how to solve this problem a bit better?

My solution:

```ql
class FunctionNodeFam extends DataFlow::FunctionNode {
    FunctionNodeFam() {this instanceof DataFlow::FunctionNode}
  
    FunctionNodeFam getAChild() {result = any(FunctionNodeFam fn | fn.getFunction().getACall().getRoot() = this.getFunction().getFuncDecl())}
 }
  
class MsgConfig extends DataFlow::Configuration {
      MsgConfig() {this = "Config of ABCI messages / transactions"}
  
      override predicate isSink(DataFlow::Node n) {
          exists(Write w |
            w.writesField(_, any(Field f | f.hasQualifiedName("google.golang.org/grpc", "MethodDesc", "Handler")), n)
          )
      }
  
      override predicate isSource(DataFlow::Node n) {
          n instanceof DataFlow::FunctionNode
      }
  }

// Final query:
from FunctionNodeFam fn, MsgQuery c
where  c.hasFlow(fn, _)
select fn.getAChild*().getFunction().getFuncDecl()

**Reply by smowton** [reply]
How about

```ql
Function reachableFromCalledService() {
  // Base case: function flows to a Handler field
  exists(Config c, DataFlow::FunctionNode serviceHandler |
    c.hasFlow(serviceHandler, _) and
    result = serviceHandler.getFunction()
  )
  or
  // Recursive case: walking the transitive call graph
  exists(DataFlow::CallNode cn |
    cn.getRoot() = reachableFromCalledService().getFuncDecl() and
    result = cn.getACalleeIncludingExternals().asFunction()
  )
}
```

It's good to restrict the base case like that because it ensures QL will only compute the subset of the transitive call graph you actually care about, not that of all functions.

**Reply by JasperSurmont** [reply]
Thank you! Very clean.

**Reply by JasperSurmont** [reply]
@smowton Sorry to bother you again. Everything still works, but I am getting very long query times because of the `reachableFromCalledService` function. Are there any improvements I could use when I'm using this function? Or are there alternative solutions that would not give such latency?

To be concrete: I'm getting query times of > 5 minutes, while the queries not using this function are under 2 seconds.

**Reply by smowton** [reply]
After enabling the preference "Code QL › Running Queries: Debug" and then running such a query on an empty cache (command: CodeQL: Clear Cache), from the output window "CodeQL Query Server", could you paste the predicate cost report beginning "Most expensive predicates for completed query..."?

**Reply by JasperSurmont** [reply]
@smowton This is what I got (showing only the top 15):
```
Most expensive predicates for stage 0 of completed query beginendblockPanic.ql:
        time  | evals |   max @ iter | predicate
        ------|-------|--------------|----------
        517ms |  1201 | 214ms @ 1    | ControlFlowGraphImpl#dd7304ee::CFG::succ#1#ff@8da47wp3
        311ms |    45 |  34ms @ 1    | ControlFlowGraphImpl#dd7304ee::CFG::ControlFlowTree::lastNode#2#dispred#fff@a4c9cwbl
        265ms |  1203 |   3ms @ 18   | #ControlFlowGraphImpl#dd7304ee::CFG::notDeferSucc#1Plus#bf#reorder_1_0@8da47yp3
         81ms |    18 |  31ms @ 2    | ControlFlowGraphImpl#dd7304ee::CFG::ControlFlowTree::firstNode#1#dispred#ff@3a865wgd
         80ms |       |              | ControlFlowGraphImpl#dd7304ee::CFG::ControlFlowTree::lastNode#2#dispred#fff_201#join_rhs@e0200d44
         74ms |       |              | AST#a1f7ddb8::AstNode::getEnclosingFunction#0#dispred#ff@c931d5q7
         67ms |    22 |   9ms @ 3    | #AST#a1f7ddb8::AstNode::parentInSameFunction#0#dispredPlus#fb#flipped@512f94u9
         63ms |       |              | exprs_1203#join_rhs@1089e2tq
         62ms |       |              | exprs_1023#join_rhs@134798mt
         61ms |       |              | exprs_1230#join_rhs@4dff5cjm
         61ms |       |              | ControlFlowGraphImpl::CFG::ControlFlowTree_not_AssignmentTree_CaseClauseTree_CompositeLiteralTree_FileTree_ForTree_FuncDefTree_IfStmtTree_IncDecTree_IndexExprTree_PostOrderTree_PreOrderTree_RangeTree_SelectStmtTree_SelectorExprTree_SendStmtTree_SliceExprTree_SwitchTree_UnaryExprTree#f@1c2ba84p
         59ms |       |              | AST#a1f7ddb8::AstNode::parentInSameFunction#0#dispred#ff#flipped@c58f83ej
         55ms |    43 |   6ms @ 4    | ControlFlowGraphImpl#dd7304ee::CFG::ControlFlowTree::lastNode#fff@a4c9czbl
         55ms |       |              | 
```

I assume the long query time is simply because of the size of the repository then? Or is there something that can be done?

**Reply by JasperSurmont** [reply]
Okay it looks like I was able to find the (or at least one) issue myself. Originally, I edited (and forgot I did) your proposed `reachableFromService` as follows (adding the sink variable):
```codeql
private FuncDecl reachableFromService(Configuration c, Node sink) {
   // Base case: function flows to a Handler field
   exists(FunctionNode serviceHandler |
       c.hasFlow(serviceHandler, sink) and
       result = serviceHandler.getFunction().getFuncDecl()
   )
   or
   // Recursive case: walking the transitive call graph
     exists(CallNode cn |
        cn.getRoot() = reachableFromService(c, sink) and
        result = cn.getACalleeIncludingExternals().asFunction().getFuncDecl()
    )
}
```

This allowed me to make path queries, because you need the sink for that. However, I then did queries like:
```codeql
predicate isInMsgPath() {
    exists(MyConfig c |
        this = reachableFromService(c, _)
    )
}
```
I think this do-not-care variable is far too unconstrained which means it will try far too many values? 
By removing it I was able to reduce the query time to a bit under 1 minute, and following queries using the similar predicates are very fast because I assume the `reachableFromService` results are cached.

**Reply by smowton** [reply]
Your pasted list of expensive predicates is the list `for stage 0` -- if you look below that you'll find a list for stage 1, stage 2 ... stage N, and then finally an overall list across all stages. Look at that last one first to find your prime suspects.

Please do also paste the complete query as it exists today.

**Reply by JasperSurmont** [reply]
Sorry, here is the correct list:

```
Most expensive predicates for completed query beginendblockPanic.ql:
        time  | evals |   max @ iter | predicate
        ------|-------|--------------|----------
        17.9s |    10 |  7.7s @ 5    | abci#efac9474::CosmosFuncDecl::reachableFromService#fbf@69caf00n
         5.2s |       |              | m#Scopes#fc940c83::Field::hasQualifiedName#3#dispred#fbbb#mcpe_f@2ab8643t
         4.5s |       |              | abci#efac9474::CosmosFuncDecl::reachableFromService#fbf_102#join_rhs@c91f06ub
         2.5s |       |              | _Packages#903ebdfa::Package::getPath#0#dispred#ff_1#join_rhs_m#ExternalFlow#8c1e37f5::interpretEleme__#shared@b4d92462
         1.6s |       |              | m#Scopes#fc940c83::Field::hasQualifiedName#3#dispred#fbbb#mcpe_tp@dee09be5
        562ms |    22 |  95ms @ 3    | #AST#a1f7ddb8::AstNode::parentInSameFunction#0#dispredPlus#fb#flipped@6b4085vm
        437ms |       |              | m#Scopes#fc940c83::Field::hasQualifiedName#3#dispred#fbbb#mcpe_pkg@d99727st
        350ms |       |              | #AST#a1f7ddb8::AstNode::parentInSameFunction#0#dispredPlus#fb#flipped_10#join_rhs@167268h5
        280ms |       |              | AST#a1f7ddb8::AstNode::getEnclosingFunction#0#dispred#fb@533389cn
        250ms |       |              | BasicBlocks#6a87fa2f::startsBB#1#f@16f87blo
```

This is the query and classes I use:
```codeql
class BlockConfig extends Configuration {
    BlockConfig() { this = "BeginBlock / EndBlock callee flow"}

    override predicate isSource(Node n) {
        n instanceof FunctionNode
    }

    override predicate isSink(Node n) {
        isBeginEndBlockFunc(n.(FunctionNode).getFunction().getName())
    }

    bindingset[s]
    private predicate isBeginEndBlockFunc(string s) {
        s.matches("%BeginBlock%") or
        s.matches("%EndBlock%")
    }
}

class CosmosFuncDecl extends FuncDecl {
    predicate isInBeginEndBlockPath() {
        exists(BlockConfig c | 
            this = reachableFromService(c)
        )
    }

   private FuncDecl reachableFromService(Configuration c) {
        // Base case: function flows to a Handler field
        exists(FunctionNode serviceHandler |
            c.hasFlow(serviceHandler, _) and
            result = serviceHandler.getFunction().getFuncDecl()
        )
        or
        // Recursive case: walking the transitive call graph
        exists(CallNode cn |
            cn.getRoot() = reachableFromService(c) and
            result = cn.getACalleeIncludingExternals().asFunction().getFuncDecl()
        )
    }
}

from CallNode cn, CosmosFuncDecl f
where
  cn.getTarget().mustPanic() and
  f = cn.getRoot() and f.isInBeginEndBlockPath()
select cn, 
  "Possible panics in BeginBock- or EndBlock-related consensus methods could cause a chain halt"

**Reply by smowton** [reply]
A few comments:

1. The dataflow configuration no longer does anything, because a FunctionNode is either a global function declaration or a function literal. They can flow to places (as function pointers), but nothing can flow to them. Therefore the only flow would be the trivial zero-length path, where source = sink.
2. Class `CosmosFuncDecl` doesn't have a characteristic predicate  (_charpred_, a constraint with constructor-like syntax), though it's clear we're only interested in declarations with `isInBeginEndBlockPath` -- therefore we can move that constraint into the charpred.
3. Predicate `reachableFromService` is declared on class `CosmosFuncDecl` but makes no reference to `this` -- as a result, it constructs the cartesian product of all `CosmosFuncDecl` (which is all `FuncDecl`) and its result (all `FuncDecl` that are reachable from some BeginBlock or EndBlock). This should be taken outside the class to prevent this cartesian product.

If we take all those together we get

```ql
import go
import DataFlow

Function getABeginEndBlockFunc() {
  result.getName().matches(["%BeginBlock%", "%EndBlock%"])
}

private FuncDecl reachableFromService() {
  // Base case: a BeginBlock or EndBlock function.
  result = getABeginEndBlockFunc().getFuncDecl()
  or
  // Recursive case: walking the transitive call graph
  exists(CallNode cn |
      cn.getRoot() = reachableFromService() and
      result = cn.getACalleeIncludingExternals().asFunction().getFuncDecl()
  )
}

class CosmosFuncDecl extends FuncDecl {
  CosmosFuncDecl() {
    this = reachableFromService()
  }
}

from CallNode cn, CosmosFuncDecl f
where
cn.getTarget().mustPanic() and
f = cn.getRoot()
select cn,
"Possible panics in BeginBock- or EndBlock-related consensus methods could cause a chain halt"
```

Of course `CosmosFuncDecl` is nothing but a synonym for `reachableFromService()`, so we could just say `cn.getRoot() = reachableFromService()` instead and drop the variable `CosmosFuncDecl f`.

I think this should be much faster.

**Reply by JasperSurmont** [reply]
Thanks! This improved the speed a lot. The reason I still had the CosmosFuncDecl is for the other data flows which we discussed earlier :smile: 

A more general question: is it good practice to have simple classes that combine different classes? Example:

```
class A extends int {
    A() {...}
}

class B extends int {
    B() {...}
}

class C extends int {
    C() {...}
}

class ABC extends int {
    ABC() {
        this instanceof A or this instanceof B or this instanceof C
    }
}
```
Is this the ideal way of creating such a generalization over `A B C`, or is there another way?

**Reply by smowton** [reply]
That pattern could be achieved like

```
abstract class ABC extends int {
 
}

class A extends ABC { ... }

class B extends ABC { ... }
```

Abstract classes are in essence the sum of their subclasses. See also https://codeql.github.com/docs/ql-language-reference/types/#abstract-classes
